api4/system.go

```diff
@@ -432,6 +434,10 @@ func getRedirectLocation(c *Context, w http.ResponseWriter, r *http.Request) {
 		w.Write([]byte(model.MapToJson(m)))
 		return
 	}
+	defer func() {
+		io.Copy(ioutil.Discard, res.Body)
```

Comment:
Potentially stupid question, but why is it necessary to read the full body before closing? What happens if we don't?

Reply:
Yes, I just got asked the same question earlier today :) It is a very common gotcha in Go, and one a lot of devs stumble upon. The answer is in the docs:

> // The http Client and Transport guarantee that Body is always
    // non-nil, even on responses without a body or responses with
    // a zero-length body. It is the caller's responsibility to
    // close Body. The default HTTP client's Transport may not
    // reuse HTTP/1.x "keep-alive" TCP connections if the Body is
    // not read to completion and closed.

https://golang.org/pkg/net/http/#Response

Reply:
Stress on "read to completion".

Reply:
We don't reuse the client, though, so I don't think keep-alive will work in either case?

Reply:
Is there a cost on reading the whole `body`?

Reply:
@hanzei If the server is slow to respond, the read will block until we hit the client's timeout -- I think. It won't matter for the end-user since we use `defer`, but the file descriptor won't be freed immediately.

Of course in this specific case a well-behaving server shouldn't return any body at all, but that's not something we can rely on.

Reply:
Yeah, we don't reuse the client, which doesn't really make sense. We should. A client is a heavy object which maintains the connections, so initializing it every time has a cost.

It is generally _highly_ recommended to read the body to completion, and there are several articles around it. But let me know if it's not necessary here.

For non-GET cases, we anyways are reading the body, so the same question about misbehaving server applies there too. This just ensures that there is no remaining bytes left to read. Most of the cases, there will be none left to read, and this will usually turn out to be a noop.

Reply:
The only potential issue I see is regarding file descriptors as mentioned above, but I don't think even that's a problem since a malicious server could just as easily attempt to serve the response headers slowly. Marking this as resolved.

Reply:
I think that is a problem very specific to this HEAD case. In the general GET case, a server may send the rest of the body also slowly. But yes, even in the HEAD case, the headers can be sent slowly. :)